identity/policies: Add support for managing policies in Triton

stack72 opened the pull request:
```
% [add-support-for-policies] % go test -v
=== RUN   TestListPolicies
=== RUN   TestListPolicies/successful
=== RUN   TestListPolicies/eof
=== RUN   TestListPolicies/bad_decode
=== RUN   TestListPolicies/error
--- PASS: TestListPolicies (0.00s)
    --- PASS: TestListPolicies/successful (0.00s)
    --- PASS: TestListPolicies/eof (0.00s)
    --- PASS: TestListPolicies/bad_decode (0.00s)
    --- PASS: TestListPolicies/error (0.00s)
=== RUN   TestGetPolicy
=== RUN   TestGetPolicy/successful
=== RUN   TestGetPolicy/eof
=== RUN   TestGetPolicy/bad_decode
=== RUN   TestGetPolicy/error
--- PASS: TestGetPolicy (0.00s)
    --- PASS: TestGetPolicy/successful (0.00s)
    --- PASS: TestGetPolicy/eof (0.00s)
    --- PASS: TestGetPolicy/bad_decode (0.00s)
    --- PASS: TestGetPolicy/error (0.00s)
=== RUN   TestDeletePolicy
=== RUN   TestDeletePolicy/successful
=== RUN   TestDeletePolicy/error
--- PASS: TestDeletePolicy (0.00s)
    --- PASS: TestDeletePolicy/successful (0.00s)
    --- PASS: TestDeletePolicy/error (0.00s)
=== RUN   TestUpdatePolicy
=== RUN   TestUpdatePolicy/successful
=== RUN   TestUpdatePolicy/error
--- PASS: TestUpdatePolicy (0.00s)
    --- PASS: TestUpdatePolicy/successful (0.00s)
    --- PASS: TestUpdatePolicy/error (0.00s)
=== RUN   TestCreatePolocy
=== RUN   TestCreatePolocy/successful
=== RUN   TestCreatePolocy/error
--- PASS: TestCreatePolocy (0.00s)
    --- PASS: TestCreatePolocy/successful (0.00s)
    --- PASS: TestCreatePolocy/error (0.00s)
=== RUN   TestListUsers
=== RUN   TestListUsers/successful
=== RUN   TestListUsers/eof
=== RUN   TestListUsers/bad_decode
=== RUN   TestListUsers/error
--- PASS: TestListUsers (0.00s)
    --- PASS: TestListUsers/successful (0.00s)
    --- PASS: TestListUsers/eof (0.00s)
    --- PASS: TestListUsers/bad_decode (0.00s)
    --- PASS: TestListUsers/error (0.00s)
=== RUN   TestGetUser
=== RUN   TestGetUser/successful
=== RUN   TestGetUser/eof
=== RUN   TestGetUser/bad_decode
=== RUN   TestGetUser/error
--- PASS: TestGetUser (0.00s)
    --- PASS: TestGetUser/successful (0.00s)
    --- PASS: TestGetUser/eof (0.00s)
    --- PASS: TestGetUser/bad_decode (0.00s)
    --- PASS: TestGetUser/error (0.00s)
=== RUN   TestDeleteUser
=== RUN   TestDeleteUser/successful
=== RUN   TestDeleteUser/error
--- PASS: TestDeleteUser (0.00s)
    --- PASS: TestDeleteUser/successful (0.00s)
    --- PASS: TestDeleteUser/error (0.00s)
=== RUN   TestCreateUser
=== RUN   TestCreateUser/successful
=== RUN   TestCreateUser/error
--- PASS: TestCreateUser (0.00s)
    --- PASS: TestCreateUser/successful (0.00s)
    --- PASS: TestCreateUser/error (0.00s)
=== RUN   TestUpdateUser
=== RUN   TestUpdateUser/successful
=== RUN   TestUpdateUser/error
--- PASS: TestUpdateUser (0.00s)
    --- PASS: TestUpdateUser/successful (0.00s)
    --- PASS: TestUpdateUser/error (0.00s)
PASS
ok  	github.com/joyent/triton-go/identity	0.012s
```

sean- commented on identity/policies.go:
```diff
@@ -0,0 +1,158 @@
+type ListPoliciesInput struct{}
+
+func (c *PoliciesClient) List(ctx context.Context, _ *ListPoliciesInput) ([]*Policy, error) {
+	path := fmt.Sprintf("/%s/policies", c.client.AccountName)
```
Not for this PR, but see #88

sean- commented on identity/policies.go:
```diff
@@ -0,0 +1,158 @@
+	"fmt"
+	"net/http"
+
+	"github.com/hashicorp/errwrap"
```
Being removed shortly...

sean- commented on identity/policies.go:
```diff
@@ -0,0 +1,158 @@
+	PolicyID string
+}
+
+func (c *PoliciesClient) Get(ctx context.Context, input *GetPolicyInput) (*Policy, error) {
```
Use PBV instead of pointer if possible

stack72 replied:
We use pointers in all other places - using a pointer here keeps us consistent

sean- commented on identity/policies.go:
```diff
@@ -0,0 +1,158 @@
+	PolicyID string
+}
+
+func (c *PoliciesClient) Delete(ctx context.Context, input *DeletePolicyInput) error {
```
PBV

stack72 replied:
We use pointers in all other places - using a pointer here keeps us consistent

sean- commented on identity/policies.go:
```diff
@@ -0,0 +1,158 @@
+	Description string   `json:"description,omitempty"`
+}
+
+func (c *PoliciesClient) Update(ctx context.Context, input *UpdatePolicyInput) (*Policy, error) {
```
PBV

stack72 replied:
We use pointers in all other places - using a pointer here keeps us consistent

sean- commented on identity/policies.go:
```diff
@@ -0,0 +1,158 @@
+	Description string   `json:"description,omitempty"`
+}
+
+func (c *PoliciesClient) Create(ctx context.Context, input *CreatePolicyInput) (*Policy, error) {
```
PBV

stack72 replied:
We use pointers in all other places - using a pointer here keeps us consistent

sean- commented on identity/policies.go:
```diff
@@ -0,0 +1,158 @@
+		Body:   input,
+	}
+	respReader, err := c.client.ExecuteRequest(ctx, reqInputs)
+	if respReader != nil {
```
We can move this to after the `err != nil`.  `ExecuteRequest()` is guaranteed to not need the response closed if an error is returned.  This makes the code more idiomatic.

sean- commented on identity/policies.go:
```diff
@@ -0,0 +1,158 @@
+		Path:   path,
+	}
+	respReader, err := c.client.ExecuteRequest(ctx, reqInputs)
+	if respReader != nil {
```
Move below `err != nil`

sean- commented on identity/policies.go:
```diff
@@ -0,0 +1,158 @@
+		Path:   path,
+	}
+	respReader, err := c.client.ExecuteRequest(ctx, reqInputs)
+	if respReader != nil {
```
Move below `err != nil`

sean- commented on identity/policies.go:
```diff
@@ -0,0 +1,158 @@
+		Path:   path,
+	}
+	respReader, err := c.client.ExecuteRequest(ctx, reqInputs)
+	if respReader != nil {
```
Move below `err != nil`

sean- commented on identity/policies.go:
```diff
@@ -0,0 +1,158 @@
+		Body:   input,
+	}
+	respReader, err := c.client.ExecuteRequest(ctx, reqInputs)
+	if respReader != nil {
```
Move down

sean- commented on identity/policies_test.go:
```diff
@@ -0,0 +1,453 @@
+)
+
+var (
+	listPoliciesErrorType = errors.New("Error executing ListPolicies request:")
```
Even though this is new and matching the style of the rest of the package, can we use this as the beachhead to start using lowercase error messages?

sean- commented on identity/policies_test.go:
```diff
@@ -0,0 +1,453 @@
+		if err != nil {
+			return nil, err
+		}
+		return policies, nil
```
Add newline above this `return`.

sean- commented on identity/policies_test.go:
```diff
@@ -0,0 +1,453 @@
+		defer testutils.DeactivateClient()
+
+		user, err := ic.Policies().Get(ctx, &identity.GetPolicyInput{
+			PolicyID: "95ca7b25-5c8f-4c1b-92da-4276f23807f3",
```
Can you add a comment explaining this magic UUID?

sean- commented on identity/policies_test.go:
```diff
@@ -0,0 +1,453 @@
+		if err != nil {
+			return nil, err
+		}
+		return user, nil
```
Add newline above `return`.

sean- commented on identity/policies_test.go:
```diff
@@ -0,0 +1,453 @@
+		defer testutils.DeactivateClient()
+
+		return ic.Policies().Delete(ctx, &identity.DeletePolicyInput{
+			PolicyID: "8700e959-4cb3-4337-8afa-fb0a53b5366e",
```
Magic constant?

sean- commented on identity/policies_test.go:
```diff
@@ -0,0 +1,453 @@
+		defer testutils.DeactivateClient()
+
+		policy, err := ic.Policies().Update(ctx, &identity.UpdatePolicyInput{
+			PolicyID:    "95ca7b25-5c8f-4c1b-92da-4276f23807f3",
```
Magic constant?

sean- commented on identity/policies_test.go:
```diff
@@ -0,0 +1,453 @@
+		if err != nil {
+			return nil, err
+		}
+		return policy, nil
```
Newline

sean- commented on identity/policies_test.go:
```diff
@@ -0,0 +1,453 @@
+    ]
+  }
+]`)
+	return &http.Response{
```
Add newline

sean- commented on identity/policies_test.go:
```diff
@@ -0,0 +1,453 @@
+    ]
+  }
+]}`)
+	return &http.Response{
```
Add `\n`

sean- commented on identity/policies_test.go:
```diff
@@ -0,0 +1,453 @@
+
+	body := strings.NewReader(`{
+    "name": "readinstance",
+    "id": "95ca7b25-5c8f-4c1b-92da-4276f23807f3",
```
Same magic UUID?  Maybe make it a `const` + comment?

sean- commented on identity/policies_test.go:
```diff
@@ -0,0 +1,453 @@
+    ]
+  }
+`)
+	return &http.Response{
```
Add `\n`

sean- commented on identity/policies_test.go:
```diff
@@ -0,0 +1,453 @@
+
+	body := strings.NewReader(`{
+    "name": "readinstance",
+    "id": "95ca7b25-5c8f-4c1b-92da-4276f23807f3",
```
UUID?

sean- commented on identity/policies_test.go:
```diff
@@ -0,0 +1,453 @@
+      "can listmachine and getmachine"
+    ],
+  }`)
+	return &http.Response{
```
`\n`

sean- commented on identity/policies_test.go:
```diff
@@ -0,0 +1,453 @@
+
+	body := strings.NewReader(`{
+  "name": "Updated Name",
+  "id": "95ca7b25-5c8f-4c1b-92da-4276f23807f3",
```
UUID?

sean- commented on identity/policies_test.go:
```diff
@@ -0,0 +1,453 @@
+
+	body := strings.NewReader(`{
+  "name": "Test Policy",
+  "id": "8700e959-4cb3-4337-8afa-fb0a53b5366e",
```
Different UUID?

stack72 commented:
@sean- this is now ready for review again, the UUIDs in the JSON just represent 2 fake policies I grabbed from the documentation. I updated all of the tests to make sure the correct UUIDs are in place

I also updated the Update test for an error when posting to a different URL the contents of another policy to show an error occurs

P.
